pymc_marketing/mmm/tvp.py

```diff
@@ -0,0 +1,148 @@
+        Time points.
+    X_mid : int or float
+        Midpoint of the time points.
+    dims : tuple of str or str
+        Dimensions of the prior. If a tuple, the first element is the name of
+        the time dimension, and the second may be any other dimension, across
+        which independent time varying priors for each coordinate are desired
+        (e.g. channels).
+    m : int
+        Number of basis functions.
+    L : int
+        Extent of basis functions. Set this to reflect the expected range of
+        in+out-of-sample data (considering that time-indices are zero-centered).
+        Default is `X_mid * 2` (identical to `c=2` in HSGP).
+    eta_lam : float
+        Exponential prior for the variance.
+    ls_mu : float
+        Mean of the inverse gamma prior for the lengthscale.
+    ls_sigma : float
+        Standard deviation of the inverse gamma prior for the lengthscale.
+    cov_func : pm.gp.cov.Covariance
+        Covariance function.
+
+    Returns
+    -------
+    pm.Deterministic
+        Time-varying prior.
+
+    References
+    ----------
+    -   Ruitort-Mayol, G., and Anderson, M., and Solin, A., and Vehtari, A. (2022). Practical
+        Hilbert Space Approximate Bayesian Gaussian Processes for Probabilistic Programming
+
+    -   Solin, A., Sarkka, S. (2019) Hilbert Space Methods for Reduced-Rank Gaussian Process
+        Regression.
+    """
+
+    if X_mid is None:
+        X_mid = float(X.mean().eval())
+    if L is None:
+        L = X_mid * 2
+
+    with pm.modelcontext(None) as model:
+        if cov_func is None:
+            eta = pm.Exponential(f"{name}_eta", lam=eta_lam)
+            ls = pm.InverseGamma(f"{name}_ls", mu=ls_mu, sigma=ls_sigma)
+            cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
+
+        model.add_coord("m", np.arange(m))  # type: ignore
+        hsgp_dims: str | tuple[str, str] = "m"
+        if isinstance(dims, tuple):
+            model.add_coord(dims[1], model.coords[dims[1]])
+            hsgp_dims = (dims[1], "m")
+
+        gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
+        phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
+        hsgp_coefs = pm.Normal(f"{name}_hsgp_coefs", dims=hsgp_dims)
+        f = phi @ (hsgp_coefs * sqrt_psd).T
+        centered_f = f - f.mean(axis=0) + 1
+        return pm.Deterministic(name, centered_f, dims=dims)
+
+
+def create_time_varying_intercept(
+    time_index: pt.sharedvar.TensorSharedVariable,
+    time_index_mid: int,
+    time_resolution: int,
+    intercept_dist: pm.Distribution,
+    model_config: dict,
+) -> pt.TensorVariable:
+    """Create time-varying intercept.
+
+    Parameters
+    ----------
+    time_index : 1d array-like of int
+        Time points.
+    time_index_mid : int
+        Midpoint of the time points.
+    time_resolution : int
+        Time resolution.
+    model_config : dict
+        Model configuration.
+    """
+
+    with pm.modelcontext(None):
```

Comment:
Some nitpick indentation comment as above. It can be assumed to be in model context because errors if more